File: base/src/test/java/com/mindera/skeletoid/threads/threadpools/ThreadPoolUtilsUnitTest.java
package com.mindera.skeletoid.threads.threadpools;

import org.junit.Before;
import org.junit.Test;

import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import static com.mindera.skeletoid.threads.threadpools.ThreadPoolUtils.mThreadTotal;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class ThreadPoolUtilsUnitTest {

    @Before
    public void beforeTest() {
        ThreadPoolUtils.mThreadTotal.set(0);
    }

    @Test(expected = UnsupportedOperationException.class)
    public void testConstructor() {
        new ThreadPoolUtils();
    }

    @Test
    public void testGetFixedThreadPool() {
        int maxThreads = 10;
        String threadPoolName = "testPool";

        ThreadPoolExecutor threadPool = ThreadPoolUtils.getFixedThreadPool(threadPoolName, maxThreads);

        assertEquals(maxThreads, threadPool.getCorePoolSize());
        assertEquals(maxThreads, threadPool.getMaximumPoolSize());
        assertEquals(0, threadPool.getKeepAliveTime(TimeUnit.MILLISECONDS));

        ThreadFactory threadFactory = threadPool.getThreadFactory();
        assertNotNull(threadFactory);
        assertTrue(threadFactory instanceof NamedThreadFactory);

        // Would be nice to check the name and maxFactoryThreads of threadFactory
    }

    @Test
    public void testGetScheduledThreadPool() {
        int maxThreads = 10;
        String threadPoolName = "testPool";

        ScheduledThreadPoolExecutor scheduledThreadPool = ThreadPoolUtils.getScheduledThreadPool(threadPoolName, maxThreads);

        assertEquals(maxThreads, scheduledThreadPool.getCorePoolSize());

        ThreadFactory threadFactory = scheduledThreadPool.getThreadFactory();
        assertNotNull(threadFactory);
        assertTrue(threadFactory instanceof NamedThreadFactory);

        // Would be nice to check the name and maxFactoryThreads of threadFactory
    }

    @Test
    public void testShutdownNull() {
        // Kind of a dummy test
        ThreadPoolUtils.shutdown(null);
    }

    @Test
    public void testShutdown() {

        ThreadPoolExecutor threadPoolExecutor = mock(ThreadPoolExecutor.class);
        when(threadPoolExecutor.getCorePoolSize()).thenReturn(5);

        ThreadPoolUtils.shutdown(threadPoolExecutor);
        assertEquals(-5, mThreadTotal.get());

        verify(threadPoolExecutor, times(1)).shutdown();
    }

    @Test
    public void testShutdownNowNull() {
        // Kind of a dummy test
        ThreadPoolUtils.shutdownNow(null);
    }

    @Test
    public void testShutdownNow() {

        ThreadPoolExecutor threadPoolExecutor = mock(ThreadPoolExecutor.class);
        when(threadPoolExecutor.getCorePoolSize()).thenReturn(5);

        ThreadPoolUtils.shutdownNow(threadPoolExecutor);
        assertEquals(-5, mThreadTotal.get());

        verify(threadPoolExecutor, times(1)).shutdownNow();
    }
}

File: base/src/test/java/com/mindera/skeletoid/threads/threadpools/ThreadPoolUtilsUnitTest.kt
package com.mindera.skeletoid.threads.threadpools

import com.mindera.skeletoid.threads.threadpools.ThreadPoolUtils.getFixedThreadPool
import com.mindera.skeletoid.threads.threadpools.ThreadPoolUtils.getScheduledThreadPool
import com.mindera.skeletoid.threads.threadpools.ThreadPoolUtils.shutdown
import com.mindera.skeletoid.threads.threadpools.ThreadPoolUtils.shutdownNow
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

class ThreadPoolUtilsUnitTest {
    @Before
    fun beforeTest() {
        ThreadPoolUtils.threadTotal.set(0)
    }

    @Test
    fun testGetFixedThreadPool() {
        val maxThreads = 10
        val threadPoolName = "testPool"
        val threadPool =
            getFixedThreadPool(threadPoolName, maxThreads)
        Assert.assertEquals(maxThreads, threadPool.corePoolSize)
        Assert.assertEquals(maxThreads, threadPool.maximumPoolSize)
        Assert.assertEquals(
            0,
            threadPool.getKeepAliveTime(TimeUnit.MILLISECONDS)
        )
        val threadFactory = threadPool.threadFactory
        Assert.assertNotNull(threadFactory)
        Assert.assertTrue(threadFactory is NamedThreadFactory)

        // Would be nice to check the name and maxFactoryThreads of threadFactory
    }

    @Test
    fun testGetScheduledThreadPool() {
        val maxThreads = 10
        val threadPoolName = "testPool"
        val scheduledThreadPool =
            getScheduledThreadPool(threadPoolName, maxThreads)
        Assert.assertEquals(maxThreads, scheduledThreadPool.corePoolSize)
        val threadFactory =
            scheduledThreadPool.threadFactory
        Assert.assertNotNull(threadFactory)
        Assert.assertTrue(threadFactory is NamedThreadFactory)

        // Would be nice to check the name and maxFactoryThreads of threadFactory
    }

    @Test
    fun testShutdownNull() {
        // Kind of a dummy test
        shutdown(Mockito.mock(ThreadPoolExecutor::class.java))
    }

    @Test
    fun testShutdown() {
        val threadPoolExecutor =
            Mockito.mock(
                ThreadPoolExecutor::class.java
            )
        Mockito.`when`(threadPoolExecutor.corePoolSize).thenReturn(5)
        shutdown(threadPoolExecutor)
        Assert.assertEquals(-5, ThreadPoolUtils.threadTotal.get())
        Mockito.verify(
            threadPoolExecutor,
            Mockito.times(1)
        ).shutdown()
    }

    @Test
    fun testShutdownNowNull() {
        // Kind of a dummy test
        shutdownNow(Mockito.mock(ThreadPoolExecutor::class.java))
    }

    @Test
    fun testShutdownNow() {
        val threadPoolExecutor =
            Mockito.mock(
                ThreadPoolExecutor::class.java
            )
        Mockito.`when`(threadPoolExecutor.corePoolSize).thenReturn(5)
        shutdownNow(threadPoolExecutor)
        Assert.assertEquals(-5, ThreadPoolUtils.threadTotal.get())
        Mockito.verify(
            threadPoolExecutor,
            Mockito.times(1)
        ).shutdownNow()
    }
}